### DISCUSSION TITLE
Editing the turns allowed for individual lanes

### DISCUSSION THREAD

**Discussion by kerbylane** [dabreegster/abstreet#929] [Q&A]
I'm trying to add a turn-only lane to an intersection. Adding a lane is easy enough. But I can't figure out how to make it turn only, instead of turn + straight. I've looked at existing turn-only lanes. But I don't see anything different about them, when compared to my new lane.

More generally, is there a detailed description of the map editing functionality. The videos available are very helpful. But I expect there are things, like turn-only lanes, that aren't covered.

**Comment by dabreegster** [comment]
Hi! Unfortunately turn lanes and turn restrictions (#70) aren't supported. If you'd be interested in working on that feature, I can help guide.

> More generally, is there a detailed description of the map editing functionality.

Also not yet. A comprehensive user guide would be really awesome, since there's so many features spread out among different tools. There are pieces of things written at https://a-b-street.github.io/docs/user/asu.html, https://github.com/a-b-street/docs/issues/2, and a few other scattered Google docs. Help also welcome here.

**Reply by dabreegster** [reply]
A very rough outline of what a new feature to designate turn lanes would need:

1) Change the UI to mark what turns are allowed. Maybe just icons indicating left/straight/right and letting the user enable/disable them? https://github.com/a-b-street/abstreet/blob/master/apps/game/src/edit/roads.rs is the lane editor UI code; it's some of the more complex because you can interact with lanes on the map or using the draggable cards, and keeping things in sync is ad-hoc.
![Screenshot from 2022-05-30 10-37-00](https://user-images.githubusercontent.com/1664407/170963953-6837ef71-57a7-4ffa-91c5-61028425c6e7.png)

2) Plumbing through a way to represent the turn lanes through the edit system. Probably just adding explicit turn lanes to `LaneSpec` would work. https://github.com/a-b-street/abstreet/tree/master/map_model/src/edits has lots of logic for saving edits in a semi-permanent format, but I think you could just piggyback on the existing `EditRoad`

3) When anything near an intersection is edited, all the turns are regenerated there using the same logic used when importing from OSM. So you'd just need to make sure https://github.com/a-b-street/abstreet/blob/master/map_model/src/make/turns.rs pays attention to lane-level turn restrictions. It should call [get_lane_level_turn_restrictions](https://github.com/a-b-street/abstreet/blob/d98f18bb186b33ff1c636412600b0c98450e1547/map_model/src/objects/lane.rs#L234) now. This parses OSM tags right now, but we could probably consider storing the allowed turn types on the `Lane` object directly. Parse from OSM once and fill those out, then let that set be edited.
